working/0110-incremental-sound-nnbd/roadmap.md

```diff
@@ -0,0 +1,874 @@
+- In general, they cannot achieve soundness and backwards compatibility, so
+clearly they need to support unsound code at least initially.  Moreover, because
+of the binary installed code base, it's not clear that they ever could get away
+from having to deal with unsoundness.
+- Some very common constructs in C# do not play with will non-nullability, and
+would result in pervasive errors in existing code.
+- They feel that some programming idioms are too painful to work around if
+  checked soundly.
+
+On the first issue, there is a massive installed base of C# code, a fair bit
+of it seems to be only available in a compiled form.  This is essentially the
+interop problem, but at a much larger scale both because of the large user base,
+and because of the large installed base.
+
+On the second issue, the main constructs that they call out are arrays allocated
+with uninitialized elements and default constructors of structs, which leave
+things uninitialized.
+
+On the third issue, they specifically call out promoting a nullably typed field
+to a non-nullable type based on a null check as an unsound pattern that they
+feel is too painful to disallow.
+
+
+The first of these is mostly not an issue for Dart at this point.
+
+On the second point, arrays are likely to be much less of an issue for Dart,
+since the structure of the `List` API encourages incremental building, whereas
+the core `C#` `Array` type does not support this.  Default constructors may be a
+place where some amount of Dart code will need to be modified, but Dart
+programming style encourages objects that are initialized in
+constructors, and provides good support for doing so.
+
+This leaves the third issue.  On the face of it, this flies in the face of the
+experience of other languages which have simply chosen to take the sound route
+on this with apparently no repercussions (Kotlin programmers seem to live fine
+with this, for example).  It's possible that their sense was that this made the
+migration story of large amounts of existing C# code easier, or that they simply
+felt that since they were unsound elsewhere, they might as well be here.  The
+migration story may come into play for Dart, but the scale of the problem is
+much smaller.
+
+#### [Java](https://blogs.oracle.com/java-platform-group/java-8s-new-type-annotations)
+
+Java provides nullability annotations that can be used by external tools to
+provide
+[sound or near-sound checking](https://checkerframework.org/manual/#nullness-checker).
+The annotations are not interpreted with the language at all.  The Java
+documentation
+[suggests](https://blogs.oracle.com/java-platform-group/java-8s-new-type-annotations) not
+relying on nullability annotations, saying "Optional Type Annotations are not a
+substitute for runtime validation".
+
+#### [Typescript](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-2-0.html)
+
+Typescript is of course unsound by design.  In addition to the core unsoundness
```

Comment:
I'd avoid phrasing this as "of course". I think this is something like "Typescript is unsound by design, as it's layered on top of JavaScripts unsound type system"?

Reply:
Shortened to "unsound by design".  I don't want go get into why.